license_manager/service/service.go

```diff
@@ -2,36 +2,295 @@ package service
+	generateRandomID = func() (string, error) {
+		id, err := uuid.NewRandom()
+		if err != nil {
+			return "", err
+		}
+		return id.String(), nil
+	}
+
+	// timeNow returns the current time, and can be stubbed out for unit tests.
+	timeNow = time.Now
+)
+
+// janitor runs in a loop to cleanup allocations and queue spots that have not
+// been refreshed in a sufficient amount of time, as well as to promote queued
+// licenses to allocations.
+func (s *Service) janitor() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	// Don't expire or promote anything during startup.
+	if s.currentState == stateStarting {
+		return
+	}
+	allocationExpiry := timeNow().Add(-s.allocationRefreshDuration)
+	queueExpiry := timeNow().Add(-s.queueRefreshDuration)
+	for _, lic := range s.licenses {
+		lic.ExpireAllocations(allocationExpiry)
+		lic.ExpireQueued(queueExpiry)
+		lic.Promote()
+	}
+}
+
+// Allocate allocates a license to the requesting invocation, or queues the
+// request if none are available. See the proto docstrings for more details.
 func (s *Service) Allocate(ctx context.Context, req *lmpb.AllocateRequest) (*lmpb.AllocateResponse, error) {
-	// If invocation is already allocated, return allocation
-	// Otherwise, get spot in queue
-	//   If request has an ID, find that spot in the queue or error
-	//   Otherwise, request goes at the back of the queue for that license
-	// Attempt to allocate available licenses to everyone in queue
-	// If not at the front of the queue, return queue spot
-	// (At the front of the queue)
-	// If license not available, return queue spot
-	// Allocate license and return allocation
-	return nil, status.Errorf(codes.Unimplemented, "Allocate() is not yet implemented")
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if len(req.GetLicenses()) != 1 {
+		return nil, status.Errorf(codes.InvalidArgument, "licenses must have exactly one license spec")
+	}
+	licenseType := formatLicenseType(req.GetLicenses()[0])
+	lic, ok := s.licenses[licenseType]
+	if !ok {
+		return nil, status.Errorf(codes.NotFound, "unknown license type: %q", licenseType)
+	}
+	invocationID := req.GetInvocationId()
+
+	if invocationID == "" {
+		// This is the first AllocationRequest by this invocation. Generate an ID
+		// and queue it.
+		var err error
+		invocationID, err = generateRandomID()
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to generate invocation_id: %v", err)
+		}
+		inv := &invocation{
+			ID:          invocationID,
+			Owner:       req.GetOwner(),
+			BuildTag:    req.GetBuildTag(),
+			LastCheckin: timeNow(),
+		}
+		lic.Enqueue(inv)
+
+		switch s.currentState {
+		case stateStarting:
+		case stateRunning:
+			lic.Promote()
+		default:
+			return nil, status.Errorf(codes.Internal, "unhandled state: %v", s.currentState)
+		}
+	}
+
+	// Invocation ID should now be known and either queued (due to the above
+	// insert, or from a previous request) or allocated (promoted from the queue
+	// by above, or asynchronously by the janitor).
```

Comment:
"promoted from the queue by above" ? where? Looks like GetAllocated won't succeed until the next janitor run? Unless it's a new allocation?

So here it will queue, and then 1second + refresh interval later it will get going? Probably ok if it only happens at restart, but from comment, it does not sound like the desired behavior?

Reply:
> "promoted from the queue by above" ? where?

Line 153 (approximately) above: `lic.Promote()`

> Looks like GetAllocated won't succeed until the next janitor run?

If this was a new invocation, the above enqueued it and then synchronously called Promote on just that license to move everything it can from the head of the queue to allocated. So it is now either successfully allocated or still queued (because there were insufficient free slots to move it from the queue)

Reply:
SGTM